pmd-core/src/main/java/net/sourceforge/pmd/util/CollectionUtil.java

```diff
@@ -532,6 +554,94 @@ public static boolean isNotEmpty(Object[] items) {
         return Collections.unmodifiableList(res);
     }
 
+    /**
+     * Map each element of the given iterable with the given function,
+     * and accumulates it into the collector.
+     */
+    public static <T, U, A, C> C map(Collector<? super U, A, ? extends C> collector,
```

Comment:
Another example of overly used generics... What is T, U, A ,C?... Maybe I'm just not that into streams...

Reply:
I think this is just "used generics" rather than overused... It's not my fault that the Collector API uses 3 type parameters...

`U, A, C` in that order correspond to the `T, A, R` type parameters of [Collector](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collector.html):
>    T - the type of input elements to the reduction operation
>    A - the mutable accumulation type of the reduction operation (often hidden as an implementation detail)
>    R - the result type of the reduction operation

On this method we have a source of `T` elements, a mapper from `T` to `U`, and a collector that accumulates a bunch of `U`s to produce a `C`, which is generally a collection of `U`.

The `A` is an implementation detail, when you make methods that *return* collectors, this is generally some wildcard to not expose it to clients. But when you *use* the collector to perform the collection operation, then you need to declare `A` to capture it. You don't need to know that to use methods that use collectors, but the implementation of the method has to capture it.

It takes some time to get used to the collector API but once you get it there's really nothing voodoo happening.

Consider that this method is super reusable, you can do `map(toSet(), listOfT, t -> t.getU())` and end up with a `Set<U>`. You can use any of the standard collectors in `Collectors`.  It's much better than writing one specialized `map` function for each kind of result type (`mapToSet`, `mapToList`, etc).